docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,509 @@
+    E.g. label "aws.alpha.kubernetes.io/volume-type" with values "sc1" and "gp2"
+    represent SDD/HDD of AWS EBS. Similarly, label
+    "failure-domain.beta.kubernetes.io/region" with values like "us-east-1d"
+    represent zone where particular AWS EBS is created.
+
+    Users will request specific storage volumes by specifying
+    `claim.Spec.Selector`, such as `region NotIn [us-east-1a, us-east-1b] AND
+    volume-type == sc1`
+
+2.  A new API group called `storage` be created to hold the a `StorageClass`
+    API resource. Storage configuration is represented as `LabelSelector`,
+    specifying parameters of volumes belonging to this class.
+
+3.  `PersistentVolume.Spec.Class` attribute is added to volumes. This attribute
+    is optional and specifies which `StorageClass` instance represents
+    storage characteristics of a particular PV.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+4.  `PersistentVolume` instances should be labelled with configuration
+    parameters that were used for their creation, such as availability zone or
+    storage type (SSD/HDD). This should be done in an admission controller.
+
+5.  `PersistentVolumeClaim.Spec.Class` attribute is added to claims. This
+    attribute specifies that only a volume with equal
+    `PersistentVolume.Spec.Class` value can satisfy a claim.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+6.  When requesting a volume with certain configuration,
+    `PersistentVolumeClaim.Spec.Selector` should be set in a way that it matches
+    only volumes with this configuration (see point 3.)
+    When both `PersistentVolumeClaim.Spec.Selector` and
+    `PersistentVolumeClaim.Spec.Class` is set, resulting volume must match both
+    selectors in the claim and in the storage class.
+
+7.  The persistent volume controller be modified to dynamically provision
+    volumes using storage classes and claim's `Selector` to create a volume that
+    has properties requested by both `PersistentVolumeClaim.Spec.Class` and
+    `PersistentVolumeClaim.Spec.Selector`.
+
+8.  The existing provisioner plugin implementations be modified to allow
+    parameterization as appropriate via storage classes and claim selectors.
+
+9.  The existing alpha dynamic provisioning feature be phased out in the
+    next release.
+
+### Controller workflow for provisioning volumes
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.
+
+    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
+        `pv.Spec.Class` are considered.
+
+    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
+        considered.
+
+    All "considered" volumes are evaluated with `claim.Spec.Selector` and the
+    smallest matching one is then bound to the claim.
```

Comment:
Implementing this in the controller/binder is fine even if the PVC can never have both...